### PR TITLE
Fix dyn pooling

### PR DESCRIPTION
**Pull request by bpickrel**
Adds support for dynamic input shape in pooling operator along with auto-padding.  This combination requires that the padding (and therefore the output shape) can't be computed until runtime.  

Padding size computations use the same formula as for the convolution op., but the inputs are somewhat different.

The first commit on this branch does the output shape computations, but onnx test fails and ref_ops_test turns up an error in computation of pooled values in padding area.  The new ref_ops_test tests in the first commit are set to look for the current output, not the correct output.  These two test fails are work in progress; dependency on [Issue 1766](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/issues/1766)

### CODE REVIEW

File: src/pad_calc.cpp [hunk 1]
```diff
@@ -111,5 +122,33 @@ shape compute_padded_shape(const shape& input,
...
 }
 
+/**
+ *   Calculate the correct output shape for a pooling with
+ *   a given input size and other parameters.  This uses
+ *   the same formula for pooling that compute_padded_shape() uses
+ *   for convolutions, but takes slightly different inputs.
+ *
+ */
+shape compute_padded_pool_shape(const shape& input,
```
Comment (CharlieL7) [inline src/pad_calc.cpp:132]: This and compute_padded_shape() look equivalent. Merge them together and rename the inputs to something that makes sense for both.

File: test/onnx/onnx_test.cpp [hunk 2]
```diff
@@ -292,16 +292,24 @@ TEST_CASE(averagepool_3d_test)
...
+            {
+                {"mode", migraphx::op::pooling_mode::average},
+                {"stride", {2, 2, 2}},
+                {"lengths", {3, 3, 3}},
+                // TODO: explicit padding is not used in auto padding scenario, but it won't
+                // default to the value the parser adds (all 0's vs all 1's).  Is this a problem?
+                {"padding", {1, 1, 1, 1, 1, 1}},
+                {"padding_mode", 2},
+            }),
+        l0);
```
Comment (CharlieL7) [inline test/onnx/onnx_test.cpp:312]: Keep the old test that did dynamic shape average pool. Create a new one that does dynamic shape auto padding.
The padding set to ones problem is from my common above in parse_pooling.cpp L187.

File: test/onnx/onnx_test.cpp [hunk 3]
```diff
@@ -3378,6 +3393,8 @@ TEST_CASE(instance_norm_type_mismatch_test)
     EXPECT(test::throws([&] { migraphx::parse_onnx("instance_norm_type_mismatch_test.onnx"); }));
 }
 
+// TEST_CASE(instance_norm_dyn_batch_test) here
+
```
Comment (CharlieL7) [inline test/onnx/onnx_test.cpp:3397]: What is this?

File: test/ref_ops_test.cpp [hunk 4]
```diff
@@ -1152,9 +1219,15 @@ TEST_CASE(conv_dyn_batch_test)
...
-                        weights);
+    mm->add_instruction(
+        migraphx::make_op("convolution",
+                          {
+                              {"padding", {1, 1}},
+                              {"stride", {2, 2}},
+                              {"padding_mode", migraphx::op::padding_mode_t::same_lower},
+                          }),
+        input,
+        weights);
```
Comment (CharlieL7) [inline test/ref_ops_test.cpp:1230]: Why was this changed

**Review by CharlieL7 — CHANGES_REQUESTED** [review]

File: src/include/migraphx/op/pooling.hpp [hunk 5]
```diff
@@ -40,10 +41,18 @@ namespace migraphx {
...
+// from 1 on up; dimensions are <batch index, channels, spatial dimensions...>
+//
 struct pooling
 {
+    //  Members mode, ceil_mode, padding_mode are all separate concepts with similar names.
     pooling_mode mode = {pooling_mode::average};
 
+    // If the input has rank other than 4 then padding, stride, lengths must all be specified
+    // since the defaults have 2-dimensions.
+
```
[the remark this answers is not in the document]
Comment (bpickrel) [reply src/include/migraphx/op/pooling.hpp:55]: I don't think it is.  It can be worked out by studying the code, but it's not standard at all in software interfaces that whether certain attributes have to be given at all is dependent on the value of a fourth input.  It's worth pointing out.

File: src/include/migraphx/op/pooling.hpp [hunk 6]
```diff
@@ -40,10 +41,18 @@ namespace migraphx {
 inline namespace MIGRAPHX_INLINE_NS {
 namespace op {
 
+// The Pooling operator mostly follows the specifications for the Onnx pooling op.
+// It assumes an NCHW layout, extended to support any number of spatial dimensions
+// from 1 on up; dimensions are <batch index, channels, spatial dimensions...>
+//
 struct pooling
 {
+    //  Members mode, ceil_mode, padding_mode are all separate concepts with similar names.
```
[the remark this answers is not in the document]
Comment (bpickrel) [reply src/include/migraphx/op/pooling.hpp:50]: I'm trying to say that these 3 class members have confusingly similar names but don't refer to the same "mode."  Is there a wording you'd prefer?

File: src/include/migraphx/op/pooling.hpp [hunk 7]
```diff
@@ -137,8 +149,20 @@ struct pooling
...
             if(padding.size() == 2 * kdims)
                 padding_factor = padding[i] + padding[i + kdims];
-            assert(input_lens[i + 2] + padding_factor >= lengths[i]);
-            std::size_t dim_size = input_lens[i + 2] + padding_factor - lengths[i];
+            std::size_t dim_size;
+            if(input_lens[i + 2] + padding_factor < lengths[i])
+            {
+                if(padding_mode == default_)
+                    MIGRAPHX_THROW(
+                        "POOLING: given padding too little for these padding window lengths");
```
[the remark this answers is not in the document]
Comment (bpickrel) [reply src/include/migraphx/op/pooling.hpp:157]: That *IS* what it means, but I'll add your suggestion.

File: src/include/migraphx/op/pooling.hpp [hunk 8]
```diff
@@ -189,6 +235,7 @@ struct pooling
                     output_dyn_dims.push_back(
                         shape::dynamic_dimension{min_spatial_dims[i], max_spatial_dims[i], {}});
                 }
+                shape out_shape(input.type(), output_dyn_dims);
```
[the remark this answers is not in the document]
Comment (bpickrel) [reply src/include/migraphx/op/pooling.hpp:238]: Removed.  I wonder why it didn't cause a compile warning.

File: test/ref_ops_test.cpp (hunk 4 above)
Comment (bpickrel) [reply test/ref_ops_test.cpp:1230]: Looks like debug code--will revert

File: src/pad_calc.cpp (hunk 1 above)
Comment (bpickrel) [reply src/pad_calc.cpp:132]: Better the way it is.  I created this function because the differences in indexing within the shapes' dimension lists made it too messy to reuse compute_padded_shape().

File: src/include/migraphx/op/common.hpp [hunk 9]
```diff
@@ -33,8 +33,9 @@ namespace migraphx {
 inline namespace MIGRAPHX_INLINE_NS {
 namespace op {
 
-// Padding mode is default_ for fixed shape padding.
-// same_lower and same_upper used for dynamic padding.
+// Specifies where to add the "extra" cell of padding if the
+// calculated padding is an odd number.  The odd cell goes at
+// the beginning of the dimension (same_lower) or end (same_upper)
```
[the remark this answers is not in the document]
Comment (bpickrel) [reply src/include/migraphx/op/common.hpp]: done

File: src/include/migraphx/op/pooling.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (bpickrel) [reply src/include/migraphx/op/pooling.hpp]: Comment has been updated

File: src/onnx/parse_pooling.cpp [hunk 10]
```diff
@@ -192,6 +172,44 @@ struct parse_pooling : op_parser<parse_pooling>
...
+            if(in_shape.dynamic())
+            {
+                // return paddings could be empty, then setting to 0 for no padding
+                // dilations (argument 4) not supported; default to all 1's
+                cal_auto_padding_size(info,
+                                      values,
+                                      values["lengths"].to_vector<std::size_t>(),
+                                      std::vector<size_t>(in_shape.ndim() - 2, 1),
+                                      in_shape.max_lens(),
+                                      paddings);
```
[the remark this answers is not in the document]
Comment (bpickrel) [reply src/onnx/parse_pooling.cpp]: Good point.  Removed.

File: test/onnx/onnx_test.cpp (hunk 2 above)
Comment (bpickrel) [reply test/onnx/onnx_test.cpp]: New test added.  Onnx parsing does not allow both the auto_pad attribute and a specified set of padding values, although the runtime code would accept it and then ignore the padding inputs.  There's also a potential inconsistency with the intent of the Onnx attribute:  auto_pad is deprecated in the Onnx spec, but internally it's used as the way we get the output size to follow the rule found in `calculate_padding` when input size is dynamic:

`int64_t output_dim     = (input_dim + stride - 1) / stride;`

I think the new test gets it right.

File: src/include/migraphx/op/pooling.hpp [hunk 11]
```diff
@@ -357,27 +406,60 @@ struct pooling
...
-        else
+        else if((padding_mode != op::padding_mode_t::default_))
         {
+            // if padding_mode is set, input was a dynamic size.  Calculate padded size now.
+
+            // wei_lens is the same as kernel_dims, but prepended with the 2 non-
+            // spatial dimensions.  For size computations, it's used like the weights
+            // tensor for convolutions.
+            std::vector<std::size_t> wei_lens;
+            wei_lens.insert(wei_lens.end(), input_lens.begin(), input_lens.begin() + 2);
```
[the remark this answers is not in the document]
Comment (bpickrel) [reply src/include/migraphx/op/pooling.hpp]: changed, although the improvement is questionable.

**Review by kahmed10 — DISMISSED** [review]
needs changes

File: test/onnx/onnx_test.cpp (hunk 3 above)
Comment (CharlieL7) [reply test/onnx/onnx_test.cpp:3471]: This is still here

**Review by CharlieL7 — APPROVED** [review]
LGTM, still have a dangling comment. I unresolved that issue

**Review by pfultz2 — APPROVED** [review]

**Comment by bpickrel** [comment]
@pfultz2, what can you tell me about Windows export macros and the Jenkins failure we're seeing?

**Comment by pfultz2** [comment]
> @pfultz2, what can you tell me about Windows export macros and the Jenkins failure we're seeing?

I am disabling the job for now. So after #2085 is merged then this should pass CI.
